design/keycloak.x/operator.md

```diff
@@ -0,0 +1,121 @@
+
+It's purpose of the upcoming Store.X initiative to provide a full fledged static configuration backend for Keycloak but there will be a mid-term preview to enable bulk imports at startup time leveraging the REST api.
+
+### Keycloak versions alignment
+
+The operator will have its release cycle aligned with Keycloak. Each operator version will support only one corresponding Keycloak version.
+
+### Upgrades
+
+In order to upgrade to a newer Keycloak version, the operator will be upgraded first to ensure full compatibility with the operand.
+
+If custom Keycloak image is not used, the operator will use a default base image. After the operator is upgraded, it automatically upgrades Keycloak too using a newer base image.
+
+
+In case a custom Keycloak image is used, the image will need to be rebuilt to perform the upgrade. This is not going to be operator's responsibility as building a custom image often requires a complex process incl. CI/CD pipelines. After the operator is upgraded, it won't manage any existing Keycloak instances until its custom image is manually rebuilt using the right Keycloak base image aligned with the operator and updated and the image coordinates are updated in the CR.
+
+### Reaugmentation process in Kubernetes
+
+We will be leveraging Kubernetes volumes to act as "caches" for the [augmented/configured](https://quarkus.io/guides/reaugmentation) version of Keycloak.
+An initial POC to show the concept has been drafted here:
+https://github.com/andreaTP/poc-mutable-jar
+
+We will use Kubernetes volumes to cache the augmented version of the binaries.
+The artifacts in the kubernetes volume will be produced by an init-container and the operation is going to result in a noop in case the volume has already been populated by a compatible augmentation.
+
+### Connecting to a database
+
+A Postgres DB will have to be provisioned externally, it's not Keycloak Operator's responsibility to manage a database. The DB credentials will be stored as K8s Secrets.
+
+In long-term plan we'll add a limited integration with a Postgres Operator to leverage its backup functionalities for Keycloak upgrades.
+
+### Observability
+
+The operator will provide CR metrics as well as it will provide integration with Prometheus, Grafana and AlertManager for both operator and operand metrics. This will be addressed in an upcoming design proposal.
+
+### Ingresses
+
+The operator will provide an out-of-the-box experience using an opinionated default Ingress configuration.
```

Comment:
That can be disabled if desired?

Reply:
one ingress that would be nice to have is an Istio Gateway + VirtualService as entry point

Reply:
@kfox1111 Added more details around the Ingress.

@hecsalazar In the initial version, we aim only at having simple opinionated Ingress/Route. We might consider extending this in the future.